pkg/datasource.go

```diff
@@ -92,18 +97,28 @@ func (o *OCIDatasource) QueryData(ctx context.Context, req *backend.QueryDataReq
 		if err != nil {
 			return nil, errors.Wrap(err, "broken environment")
 		}
-		metricsClient, err := monitoring.NewMonitoringClientWithConfigurationProvider(configProvider)
-		if err != nil {
-			return nil, errors.New(fmt.Sprint("error with client", spew.Sdump(configProvider), err.Error()))
-		}
-		identityClient, err := identity.NewIdentityClientWithConfigurationProvider(configProvider)
-		if err != nil {
-			o.logger.Error("error with client")
-			panic(err)
+		if configProvider != nil {
+			metricsClient, err := monitoring.NewMonitoringClientWithConfigurationProvider(configProvider)
+			if err != nil {
+				return nil, errors.New(fmt.Sprint("error with client", spew.Sdump(configProvider), err.Error()))
+			}
+			identityClient, err := identity.NewIdentityClientWithConfigurationProvider(configProvider)
+			if err != nil {
+				o.logger.Error("error with client")
+				panic(err)
+			}
+			o.identityClient = identityClient
```

Comment:
I am not really sure I understand which approach you would like here. If I understood correctly here you would like to have an array of what (or partially what) is contained in OCIDataSource, let say in a new struct called OCIConfigurationSets, basically capturing  identityClient, metricsClient etc, but for each of the tenancies found in the config file. Did I understood correctly ? If so, how then I could make use of this array to switch context in the queries ? And how this approach can coexist to the current single tenancy approach, which will set the context at the very beginning of the plugin configuration ? Can you provide an example or explain a bit more on how to make use of this new function/struct ?

Reply:
Maybe one solution could be to substitute the *OCIDatasource receiver with an array of o *OCIDatasource-s, but again, I am not sure this can be easily implemented. Any thoughts ?

Reply:
My thought would be to have:
- A data structure such as:
```
type TenancyAccess struct {
	metricsClient    monitoring.MonitoringClient
	identityClient   identity.IdentityClient
	config           common.ConfigurationProvider
}
```
- Update the OCIDatasource data structure to be along the lines of:
```
type OCIDatasource struct {
	tenancyAccess map[string]*TenancyAccess
	logger           log.Logger
	nameToOCID       map[string]string
	timeCacheUpdated time.Time
}
```
where the key for the `tenancyAccess` hash map would be the `tenancyConfig` string for each tenancy
- In `QueryData()` instead of checking if o.config is nil the function would retrieve the hash map entry for the specified tenancyConfig and if it doesn't exist then the `tenancyAccess` map would be populated based on the contents of the OCI config file.
- The retrieved `TenancyAccess` structure would be passed to each of the response functions called by `QueryData()` and the information in the structure would be used to fulfill the necessary operation.
- For the single tenancy configuration, I would think there could just be a single entry in the hash map with a key of 'DEFAULT/<tenancy OCID>'.

Reply:
This approach is now implemented, but need to be carefully tested. single and multitenancy are working, variables for single mode seem to work properly, variables in multitenancy mode are not working at this stage.

Reply:
It works but only with the workaround we discussed, for now. A separate story will be filed to find out the correct condition to make it works as expected.